### PR TITLE
Introduce an admin REST API to import tenant theme

### PR DESCRIPTION
**Pull request by hisanhunais**
### Purpose
This PR will introduce an admin REST API to import tenant theme.

### Approach
1. Added a new resource **/tenant-theme/import** to admin-api.yaml which requires a zip archive containing theme configuration.
2. Introduced a new scope **apim:tenant_theme_manage** and assigned it to the /tenant-theme/import resource
3. Introduced a new method called ***deplyTenantTheme** in **RestApiAdminUtils.java** class which handles the tenant theme importing function.

### User Stories
1. Scenario 1 : Normal Flow
- Request
```
curl -k -F "file=@sample-theme.zip" -X POST -H "Authorization: Bearer d886cb32-b9db-316a-a44e-0811eef76306" https://localhost:9443/api/am/admin/v1/tenant-theme/import
```
- Response
```
Theme imported successfully
```
2. Scenario 2 : When zip archive contains an unsupported file
- Request
```
curl -k -F "file=@sample-theme-unsupported.zip" -X POST -H "Authorization: Bearer d886cb32-b9db-316a-a44e-0811eef76306" https://localhost:9443/api/am/admin/v1/tenant-theme/import
```
- Response
```
{"code":500,"message":"Internal server error","description":"Error while importing tenant theme for tenant carbon.super","moreInfo":"","error":[]}
```
- Carbon log snippet
```
[2020-05-30 15:37:51,640] ERROR - APIUtil Unsupported file is uploaded with tenant theme by carbon.super : file name : apim/css/custom.cssd
[2020-05-30 15:37:51,645] ERROR - TenantThemeApiServiceImpl Error while importing tenant theme for tenant carbon.super
org.wso2.carbon.apimgt.api.APIManagementException: Unsupported file is uploaded with tenant theme by carbon.super : file name : apim/css/custom.cssd
```

### Test Environment
JDK 1.8.0_241
Ubuntu 18.04.4 LTS

Fixes: https://github.com/wso2/product-apim/issues/8114

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/resources/admin-api.yaml [hunk 1]
```diff
@@ -2965,6 +2966,64 @@ paths:
           schema:
             $ref: '#/definitions/Error'
 
+######################################################
+# The Tenant Theme resource APIs
+######################################################
+  /import/tenant-theme:
+
```
Comment (tgtshanika) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/resources/admin-api.yaml:2973]: tenant theme is the resource here. Hence shall we use tenant-theme/upload or import?

Comment (hisanhunais) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/resources/admin-api.yaml:2973]: Fixed in https://github.com/wso2/carbon-apimgt/pull/8595/commits/fb8159849c21f633958ff21963a55d40da2ca155

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/TenantThemeImportManager.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                            fileOutputStream.write(buffer, 0, len);
+                        }
+
+                        fileOutputStream.close();
+                    } else {
+                        handleException(
+                                "Unsupported file is uploaded with tenant theme by " + tenantDomain + " : file name : "
+                                        + zipEntry.getName());
+                    }
+
```
Comment (tgtshanika) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/TenantThemeImportManager.java:116]: remove new line

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/TenantThemeImportManager.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        } catch (IOException ex) {
+            handleException("Failed to deploy tenant theme", ex);
+        } finally {
+            IOUtils.closeQuietly(zipInputStream);
+            IOUtils.closeQuietly(themeFile);
+        }
+    }
+
+    private static void handleException(String msg) throws APIManagementException {
+
```
Comment (tgtshanika) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/TenantThemeImportManager.java:133]: check whether we can reuse org.wso2.carbon.apimgt.impl.utils.APIUtil.handleException;

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/resources/admin-api.yaml [hunk 4]
```diff
@@ -2965,6 +2966,64 @@ paths:
...
+        This operation can be used to import a DevPortal tenant theme.
+      parameters:
+        - name: file
+          in: formData
+          description: |
+            Zip archive consisting of tenant theme configuration
+          required: true
+          type: file
+        - name: tenantDomain
+          in: query
```
Comment (tgtshanika) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/resources/admin-api.yaml:2998]: Do we really have to provide the tenant domain name? Shouldn't we extract the tenant domain from the auth context? Unless an unauthorized tenant user can update the tenant theme of some other tenant domain.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/RestApiAdminUtils.java [hunk 5]
```diff
@@ -55,4 +73,89 @@ public static boolean isBlockConditionAccessibleToUser(String user, BlockConditi
...
+                FileUtils.cleanDirectory(folder);
+            }
+
+            //get the zip file content
+            zipInputStream = new ZipInputStream(themeFile);
+            //get the zipped file list entry
+            ZipEntry zipEntry = zipInputStream.getNextEntry();
+
+            while (zipEntry != null) {
+
```
Comment (vithu30) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/RestApiAdminUtils.java:111]: Shall we remove the empty lines. Please check other places as well.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/RestApiAdminUtils.java [hunk 6]
```diff
@@ -55,4 +73,89 @@ public static boolean isBlockConditionAccessibleToUser(String user, BlockConditi
...
+                        }
+                    }
+                } else {
+                    String ext = FilenameUtils.getExtension(zipEntry.getName());
+                    if (EXTENSION_WHITELIST.contains(ext)) {
+                        //create all non exists folders
+                        //else you will hit FileNotFoundException for compressed folder
+                        new File(newFile.getParent()).mkdirs();
+                        FileOutputStream fileOutputStream = new FileOutputStream(newFile);
+
```
Comment (kavishkafernando) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/RestApiAdminUtils.java:137]: shall we remove this extra line?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/RestApiAdminUtils.java (hunk 5 above)
Comment (hisanhunais) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/RestApiAdminUtils.java:111]: Fixed with https://github.com/wso2/carbon-apimgt/pull/9702/commits/e22c474785b0d7f5a6c4c0ded773ac0ad2d03fc5

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/RestApiAdminUtils.java (hunk 6 above)
Comment (hisanhunais) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/RestApiAdminUtils.java:137]: Fixed with https://github.com/wso2/carbon-apimgt/pull/9702/commits/e22c474785b0d7f5a6c4c0ded773ac0ad2d03fc5

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/resources/admin-api.yaml (hunk 4 above)
Comment (hisanhunais) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/resources/admin-api.yaml]: Fixed with https://github.com/wso2/carbon-apimgt/pull/8595/commits/6ae0bc73c2aefce667fc3908529e7750aaa25d97

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/TenantThemeImportManager.java (hunk 2 above)
Comment (hisanhunais) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/TenantThemeImportManager.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/8595/commits/6ae0bc73c2aefce667fc3908529e7750aaa25d97

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/TenantThemeImportManager.java (hunk 3 above)
Comment (hisanhunais) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/TenantThemeImportManager.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/8595/commits/6ae0bc73c2aefce667fc3908529e7750aaa25d97

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/TenantThemeImportManager.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+                }
+                zipEntry = zipInputStream.getNextEntry();
+            }
+
+            zipInputStream.closeEntry();
+            zipInputStream.close();
+
+        } catch (IOException ex) {
+            handleException("Failed to deploy tenant theme", ex);
```
[the remark this answers is not in the document]
Comment (hisanhunais) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/TenantThemeImportManager.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/8595/commits/6ae0bc73c2aefce667fc3908529e7750aaa25d97

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/TenantThemeImportManager.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        byte[] buffer = new byte[1024];
+
+        String outputFolder = getStoreTenantThemesPath() + tenantDomain;
+
+        try {
+            //create output directory if it does not exist
+            File folder = new File(outputFolder);
+            if (!folder.exists()) {
+                if (!folder.mkdirs()) {
+                    handleException("Unable to create tenant theme directory");
```
[the remark this answers is not in the document]
Comment (hisanhunais) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/admin/v1/utils/TenantThemeImportManager.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/8595/commits/6ae0bc73c2aefce667fc3908529e7750aaa25d97

**Pull request merged**